sdk/core/azure-core/src/main/java/com/azure/core/util/implementation/BinaryDataContent.java

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.util.implementation;
+
+import com.azure.core.util.BinaryData;
+import com.azure.core.util.serializer.JsonSerializer;
+import com.azure.core.util.serializer.JsonSerializerProviders;
+import com.azure.core.util.serializer.ObjectSerializer;
+import com.azure.core.util.serializer.TypeReference;
+import reactor.core.publisher.Flux;
+
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.ReadOnlyBufferException;
+
+/**
+ * An abstract internal representation of the content stored in {@link BinaryData}.
+ */
+public abstract class BinaryDataContent {
+
+    static final byte[] ZERO_BYTE_ARRAY = new byte[0];
+    static final int STREAM_READ_SIZE = 1024;
+    static final JsonSerializer SERIALIZER = JsonSerializerProviders.createInstance(true);
+
+    /**
+     * Gets the length of the {@link BinaryDataContent} if it is able to be calculated.
+     * <p>
+     * If the content length isn't able to be calculated null will be returned.
+     * @return The length of the {@link BinaryDataContent} if it is able to be calculated, otherwise null.
+     */
+    public abstract Long getLength();
+
+    /**
+     * Returns a {@link String} representation of this {@link BinaryDataContent} by converting its data using the UTF-8
+     * character set.
+     * @return A {@link String} representing this {@link BinaryDataContent}.
+     */
+    public abstract String toString();
+
+    /**
+     * Returns a byte array representation of this {@link BinaryDataContent}.
+     * @return A byte array representing this {@link BinaryDataContent}.
+     */
+    public abstract byte[] toBytes();
+
+    /**
+     * Returns an {@link Object} representation of this {@link BinaryDataContent} by deserializing its data using the
+     * {@link JsonSerializer}.
+     *
+     * <p><strong>Get a non-generic Object from the BinaryDataContent</strong></p>
+     * @param <T> Type of the deserialized Object.
+     * @param typeReference The {@link Class} representing the Object's type.
+     * @return An {@link Object} representing the JSON deserialized {@link BinaryDataContent}.
+     * @throws NullPointerException If {@code typeReference} is null.
+     * @see JsonSerializer
+     */
+    public abstract <T> T toObject(TypeReference<T> typeReference, ObjectSerializer serializer);
```

Comment:
Could this have a default implementation?

Reply:
We could but wanted to leave this as all abstract methods so that the extending classes can choose the best way to implement this as it forces all the subtypes to think about how best to convert to object. If we find that there's a lot of duplication, we can always include a default impl later as this is an implementation type.